lightning/src/util/macro_logger.rs

```diff
@@ -155,12 +155,17 @@ macro_rules! log_spendable {
 	}
 }
 
+/// Create a new Record and log it. You probably don't want to use this macro directly,
+/// but it needs to be exported so `log_trace` etc can use it in external crates.
+#[macro_export]
 macro_rules! log_internal {
 	($logger: expr, $lvl:expr, $($arg:tt)+) => (
-		$logger.log(&::util::logger::Record::new($lvl, format_args!($($arg)+), module_path!(), file!(), line!()));
+		$logger.log(&$crate::util::logger::Record::new($lvl, format_args!($($arg)+), module_path!(), file!(), line!()));
```

Comment:
Wait, how does this work? is `crate` resolved with reference to the place the macro is defined? I guess if it works, it works, but that surprises me.

Reply:
I don't understand it 100% but I think it's the special magic explained here: http://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/macros.html#the-variable-crate